include/nbl/video/IGPURayTracingPipeline.h

```diff
@@ -0,0 +1,107 @@
+                EARLY_RETURN_ON_FAILURE = base_flag(EARLY_RETURN_ON_FAILURE),
+                LINK_TIME_OPTIMIZATION = base_flag(LINK_TIME_OPTIMIZATION),
+                RETAIN_LINK_TIME_OPTIMIZATION_INFO_BIT_EXT = base_flag(RETAIN_LINK_TIME_OPTIMIZATION_INFO),
+                SKIP_TRIANGLES = base_flag(SKIP_TRIANGLES),
+                SKIP_AABBS = base_flag(SKIP_AABBS),
+                NO_NULL_ANY_HIT_SHADERS = 1<<14,
+                NO_NULL_CLOSEST_HIT_SHADERS = 1<<15,
+                NO_NULL_MISS_SHADERS = 1<<16,
+                NO_NULL_INTERSECTION_SHADERS = 1<<17,
+                ALLOW_MOTION = 1<<20,
+            };
+            #undef base_flag
+
+            inline SSpecializationValidationResult valid() const
+            {
+                if (!layout)
+                  return {};
+
+                SSpecializationValidationResult retval = {
+                    .count=0,
+                    .dataSize=0,
+                };
+                const bool valid = pipeline_t::SCreationParams::impl_valid([&retval](const IGPUShader::SSpecInfo& info)->bool
+                {
+                    const auto dataSize = info.valid();
+                    if (dataSize<0)
+                        return false;
+                    else if (dataSize==0)
+                        return true;
+
+                    const size_t count = info.entries ? info.entries->size():0x80000000ull;
+                    if (count>0x7fffffff)
+                        return {};
+                    retval += {.count=dataSize ? static_cast<uint32_t>(count):0,.dataSize=static_cast<uint32_t>(dataSize)};
+                    return retval;
+                });
+                if (!valid)
+                    return {};
+                return retval;
+            }
+
+            inline std::span<const IGPUShader::SSpecInfo> getShaders() const { return shaders; }
+
+            // TODO: Could guess the required flags from SPIR-V introspection of declared caps
+            core::bitflag<FLAGS> flags = FLAGS::NONE;
+        };
+
+        inline core::bitflag<SCreationParams::FLAGS> getCreationFlags() const { return m_flags; }
+
+        // Vulkan: const VkPipeline*
+        virtual const void* getNativeHandle() const = 0;
+
+        virtual const SShaderGroupHandle& getRaygen() const = 0;
+        virtual const SShaderGroupHandle& getMiss(uint32_t index) const = 0;
+        virtual const SShaderGroupHandle& getHit(uint32_t index) const = 0;
+        virtual const SShaderGroupHandle& getCallable(uint32_t index) const = 0;
+
+    protected:
+        IGPURayTracingPipeline(const SCreationParams& params) : IBackendObject(core::smart_refctd_ptr<const ILogicalDevice>(params.layout->getOriginDevice())),
+            pipeline_t(params),
+            m_flags(params.flags)
+        {}
+
+        virtual ~IGPURayTracingPipeline() = default;
+
+        const core::bitflag<SCreationParams::FLAGS> m_flags;
```

Comment:
can you compute the default stack size https://registry.khronos.org/vulkan/specs/latest/html/vkspec.html#ray-tracing-pipeline-stack
and keep it here + provide a getter for it?

Also this way, even for dynamic stack pipelines we can validate that the stack value people are setting later on via the command is between just the raygen stack size and the stack size determined here (inclusive)

Reply:
Also keep a stack size (in uint16_t) per shader group, for the hit group you need a struct with 3 values